### PR TITLE
Extend 'HashFn' to support MD5 and be truly extensible

### PR DESCRIPTION
**Pull request by hvr**
NB: This requires a serious version bump!

### CODE REVIEW

**Comment by dcoutts** [comment]
@edsko I'm not totally clear on the consequences of adding the fallthrough/other case, but it's clear we need to be able to handle (if only by ignoring) new unknown hash formats, otherwise as soon as we add sha512 or whatever in future it'll break all deployed clients. And these hashes are supposed to be extensible.

File: hackage-security/src/Hackage/Security/TUF/FileInfo.hs [hunk 1]
```diff
@@ -87,10 +89,13 @@ fileInfoSHA256 FileInfo{..} = Map.lookup HashFnSHA256 fileInfoHashes
 
 instance Monad m => ToObjectKey m HashFn where
   toObjectKey HashFnSHA256 = return "sha256"
+  toObjectKey HashFnMD5    = return "md5"
+  toObjectKey (HashFnUnknown s) = return s
 
 instance ReportSchemaErrors m => FromObjectKey m HashFn where
   fromObjectKey "sha256" = return HashFnSHA256
-  fromObjectKey str      = expected "valid hash function" (Just str)
```
Comment (hvr) [inline hackage-security/src/Hackage/Security/TUF/FileInfo.hs]: NB: this line is what would currently cause the instance for `fileInfoHashes :: Map HashFn Hash` to choke on unknown keys, if `package.json` files would start to provide an `md5` hash as well...

**Comment by hvr** [comment]
As far as `hackage-security`'s code is concerned, extending `HashFn` doesn't seems to of little consequence; it's only ever used in `FileInfo.hs`:

```
$ git grep HashFnSHA256
hackage-security/src/Hackage/Security/TUF/FileInfo.hs:data HashFn = HashFnSHA256
hackage-security/src/Hackage/Security/TUF/FileInfo.hs:          (HashFnSHA256, Hash $ BS.C8.unpack $ Base16.encode $ SHA256.hashlazy bs)
hackage-security/src/Hackage/Security/TUF/FileInfo.hs:fileInfoSHA256 FileInfo{..} = Map.lookup HashFnSHA256 fileInfoHashes
hackage-security/src/Hackage/Security/TUF/FileInfo.hs:  toObjectKey HashFnSHA256 = return "sha256"
hackage-security/src/Hackage/Security/TUF/FileInfo.hs:  fromObjectKey "sha256" = return HashFnSHA256
```

**Comment by dcoutts** [comment]
So @edsko and I had a chat about this. He thinks the better approach for unknown map keys is to ignore them, rather than parse them as unknown, since this is consistent with the way we do things elsewhere. So that's what I'm trying in #163.

**Pull request closed without merging**